### ISSUE TITLE
[RFC] template based routing

### ISSUE THREAD

**Issue by bigopon** [aurelia/aurelia#2038]
## 💬 RFC

Use html instead of JS/TS to express routing

### 1. 🔦 Context

Currently, there is not a way to declare our routing configuration where the corresponding viewport is, consider a way to support this, so that we can have a more declarative routing.

Consider the following example:
- an `<app>` component with template
    ```html
    <h1>Items management</h1>
    <au-viewport></au-viewport>
    ```
- a reusable `item-editor` for list/detail component for editing items in a list
- 2 routes:
    - `/store/:storeId` for editing item in a specific store
    - `/warehouse` for editing item in the central warehouse
- With current routing configuration, we will have something like this:
    - separate components for separate routes:
        ```ts
        routes = [
          { path: '/store/:storeId', component: StoreEditor },
          { path: '/warehouse', component: WarehouseEditor },
        ]
        ```
    - Shared component for both routes
        ```ts
        routes = [
              { path: '/store/:storeId', component: StorageEditor },
              { path: '/warehouse', component: StorageEditor },
        ]
        ```
        and inside the component we can retrieve the route/route data option to display the ui accordingly.

In both cases above:
- the route configs and viewport declarations `<au-viewport>`) are far from each other, making it kind of "scattered"
- it's not a direct relationship between the routable components (either `StoreEditor`/`WarehouseEditor`) and the `<app>` component, making it difficult to bind directly between the two.
- routing always involves JS/TS code, even for simple routes

It may be proper to consider a different way to support routing that can provide some improvements for the above points.

### 2. 💻 Examples

1. Simple usage
    For the above examples, we can think of a html based routing inside the `<app>` component:
    
    ```html
    <h1>Items management</h1>
    <au-viewport path="/store/:storeId">
        <h2>Store ${$path.storeId | storeName} items</h2>
        <item-editor items.bind="state.storeItems[$path.storeId]"></item-editor>
    </au-viewport>
    <au-viewport path="/warehouse">
        <h2>Warehouse items</h2>
        <item-editor items.bind="state.warehouseItems"></item-editor>
    </au-viewport>
    ```
    
    There are a few notable differences:
    - there's no need for JS or TS to declare our routing
    - the viewports (`<au-viewport>`) contain the routing information (routing "table")
    - the template of the component of the matching route is declared in the app template, no wrapper components needed for our `<item-editor>`
    - `<item-editor>` components are "part" of the `<app>` component, and thus, can be bound directly with `<app>`

2. async components
    Because routing template (render) is declared in the template of the owning component, one can declare async component for a route using `<au-compose>`, like the following example
    ```html
    <h1>Items management</h1>
    <au-viewport path="/store/:storeId">
        <h2>Store ${$path.storeId | storeName} items</h2>
        <item-editor items.bind="state.storeItems[$path.storeId]"></item-editor>
        <au-compose component.bind="getStoreMonitor()"></au-compose>
    </au-viewport>
    <au-viewport path="/warehouse">
        <h2>Warehouse items</h2>
        <item-editor items.bind="state.warehouseItems"></item-editor>
        <au-compose component.bind="getItemModelViewer()"></au-compose>
    </au-viewport>
    ```

3. route nesting
    With standard JS based routing, we can declare `<au-viewport>` inside a routable component, which are `StoreEditor`/`WarehouseEditor` in our original example above. For template base routing declaration, we can have one `<au-viewport>` nested inside another, like the following example:
    ```html
    <h1>Items management</h1>
    <au-viewport path="/store">
        <au-viewport path="/">
          All stores index view
        </au-viewport>

        <au-viewport path="/:storeId">
          <h2>Store ${$path.storeId | storeName} items</h2>
          Store items view
        </au-viewport>
    </au-viewport>
    ```

4. route guarding
    Because routing no longer involves a component, we need a different way to specify route guard. Guarding can be a bindable on the `<au-viewport>` itself, like the following example:
    ```html
    <h1>Items management</h1>
    <au-viewport path="/store">...</au-viewport>
    
    <au-viewport path="/warehouse" guard.bind="(path, params, ...etc) => ...">
    <au-viewport path="/warehouse" guard.bind="state.isWarehouseManager">
    ```

5. redirecting
    A redirect API on the `<au-viewport>` itself, like the following example:
    ```html
    <h1>Items management</h1>
    <au-viewport path="/store">
        ...
    </au-viewport>

    <au-viewport path="*" redirect="/store"></au-viewport>
    ```
    Redirecting from the component view model instances can probably be done through injecting a current route context and call navigate from that context.

6. route config
    Because each route rendering is part of a template, config to each route template can be standard binding(s), so there' not a need for routing config with the existing modes

### 3. 🖼️ Prior arts

- React router allows applications to declare routing in similar manner above
- maybe others as well, however I'm not familiar

### 4. ⌨️ Implementation

If we are to implement this, should it be incorporated into the existing routers, or be a new separate router.
- If we are to incorporate this routing into existing routers, how it will play with the existing modes? It can get potentially confusing. One way to incorporate is to have a new "viewport" component:
    ```html
    <au-route path="...">
      ...
    </au-route>
    ```
- If we are to create a new router, what do we need to consider?

### 5. 💭Evaluation

Template based routing declaration is a good and simple way to design the structure of an application.
- It allows routes to be part of a component template, thus application in obvious way, instead of a separate realm/regions inside an app.
- It leverages Aurelia power in the template and allow more work to be done simpler before having to move code to a JS/TS component.

**Comment by Sayan751** [comment]
@bigopon Thank you for the detailed RFC. The examples are helpful for understanding the intention.

Playing the role of a critic, I would like to mention a couple of things that I found a bit odd.

### 1. The emphasis on avoiding JS/TS in favour of HTML-only components

I think that it is generally true, if I say, that both JS/TS and HTML offers flexibility and strength in certain areas, and thus, it is for the best when those respective strengths are leveraged. I am of the opinion that when the routing configuration is done in the JS/TS it is more concise than replicating that in HTML. For example, here is one of the simplest route configuration in a typical app at work.

```js
{
  id: "foo",
  path: "",
  title: "app.routes.r.foo", //<-- i18n key for the title
  component: "foo-list",
  nav: true,                 //<-- for the menu renderer CE; if this route should be visible in the menu
  data: {
    claim: {                 // <-- for the Auth hook; whether the user is allowed to navigate to the route
      required: "all",
      items: [
        { resource: "foo", type: "read" },
      ]
    }
  }
}
```
It seems to me that accommodating all that in HTML might produce more verbose code than warranted. One may choose to define the route-data in JS/TS, because the `data` has no schema and considered more-or-less a bag of properties, the `au-viewport` or `au-route` can only accommodate a generic bindable for the `data` property (instead of having separate bindable for `claim` etc.). And once you do so, one is scattering the routing information between JS/TS and HTML. In that case, it makes more sense to consolidate everything in JS/TS.

Speaking of JS/TS, even some of your examples demonstrates usage of view-model properties for route guard and conditionally rendering a route. So one cannot do without the JS/TS completely, and some routing configuration/setting will still come from JS/TS.

Hence, in conclusion, I would like to express that the emphasis on avoiding JS/TS for routing is somewhat incomprehensible to me.

### 2. Leaky encapsulation

I am of this opinion that this proposal somehow encourages leaky encapsulation. This is elaborated in the following subsections.

#### 2.1. Binding to routed-components

This is also a bit counter-intuitive to me. I am used to it that the routed components are self-sufficient enough to get their own-data using services etc., using only the parameters from the route. The idea of binding data from a host-component (for simplicity of the discussion, let us refer that as app-root) to the routed-components gives me the idea that those routed components are not self-sufficient. In other words, the encapsulation is not tight enough. This might also encourage the app-root to be a mishmash of logic, meant for various other routed-components.

#### 2.2. Class-less routed-components

It is also true that this pattern can encourage class-less routed-components, as the template is defined in the app-root. From the initial look, it is very much tempting; especially for trivial routed-components. However, according to my experience, I never encountered a routed-component so trivial. Therefore, it seems to me that it is very much likely, that for a real-world application, the HTML for the app-root is going to be a wall of code.

Wrapping up this point, I would like to underscore that in my opinion, the responsibility of the app-root ends with defining a routing table/configuration and one or more viewports where the routed-components will be loaded; the responsibilities of how those components will get data, or how the view will look like, are delegated to the components.

P.S. I didn't quite understand the use-case of using `au-compose` for async components, and why this new pattern is helpful for that, or why the current practice hinders that.

### 3. Guarding

The term "guard" is a bit ambiguous to me. In general, it says to me that it determines whether a component is allowed to be loaded or not. This can be done in several ways:

1. **Employ a `canLoad` hook.** If you have meant that, then I would like to point out that that adds one more point for the leaky encapsulation 😛 
2. **Employ a [route-parameter constraint](https://docs.aurelia.io/getting-to-know-aurelia/routing/aurelia-router-lite/configuring-routes#constrained-parameters).** If you have meant that, then I would say that it is best to use that directly with the route-parameters.
3. **Employ a [router hook](https://docs.aurelia.io/getting-to-know-aurelia/routing/aurelia-router-lite/router-hooks).** For a declarative approach, it makes sense to specify route specific hooks. However, as the hooks are capable of more than just guarding the routes, probably we need a more non-ambiguous name for that, in case you meant that. 

### 4. Dynamic/Conditional routes

I think that instead of using `if.bind` we can stick to the router-hooks, as those are actually meant for this purpose. It seems to me like reinventing the wheel.

### 5. Integration with configured routing

The RFC does not mention how the new pattern with play with the current practice. Think of a hybrid app, using configured routing (current practice) as well as the declarative routing.

### 6. Testing

This is sort of a continuation of the 2nd point. Due to the leaky encapsulation, it seems to me that testing a component in isolation will be harder in this pattern.

Those are all the critiques I have for this proposal at this moment. Note that the choice of a design pattern is very much subjective, and may vary from project to project or from one application to another. Even though I have so far no need for declarative routing, does not mean we should not have that. I just wanted to mention the things that seemed problematic to me.

For the implementation part, I think ideally we should have something like a router-plugin that implements the new pattern and can be used for both routers. And we should be careful not to name it like a _third_ router; I think despite all the differences, everyone is likely to agree on that 😉

**Comment by bigopon** [comment]
Thanks @Sayan751 , there' quite a lot of comments I may miss one or two

- For the nav configuration
    >   nav: true,                 //<-- for the menu renderer CE; if this route should be visible in the menu

    I don't see why it's an issue with this config, one can have
    ```html
    <au-viewport nav="true">
    ```
    and it will just be as easy to reflect that on the nav model

- For auth configuration
    ```
    data: {
       claim: {                 // <-- for the Auth hook; whether the user is allowed to navigate to the route
    ```
    it's a choice. One can split the routing auth to each route, in all places where there's a route configuration, or they can also do it in a single place, or the root hook. Pseudo code for discussion purpose:
    ```ts
    class RouterHook {
      canLoad(path, viewport) {
        viewport.data // for auth purposes here if one prefers
      }
    }
    ```
    and template:
    ```html
    <au-viewport data.bind="...">
    ```
    This is even better since my viewport now has all the information I want to see in one place:
    ```html
    <au-viewport path="..." nav="true" data.bind="...">
    ```
    I personally wouldn't mind this, especially for small enough scale app I'd prefer to just do this. 

- For the binding to `routed-components`
    > ... I am used to it that the routed components are self-sufficient enough to get their own-data using services etc.
    
    To me, the removal of unnecessary encapsulation is the goal. I don't think in my examples, there's a need for an encapsulating routed-component, at all. If we were allowed to say `I'm used to ..,`, I'd say `I'm used to plain and simple template with fewer to no indirections`

    > ... those routed components are not self-sufficient ...
    > ... This might also encourage the app-root to be a mishmash of logic, meant for various other routed-components.

    May I know sufficient in which area? If we are talking about routing logic, maybe another perspective is to ask whether we want to scatter the routing logic all over the places? Maybe logic of routing should be mishmash-ed into the app component, at least until it's not suitable. Also, we can always have a page component to handle landing to maintain the routing logic split, if preferred:
    ```html
    <au-viewport path="/store">
      <my-store-page>
    </au-viewport>
    ```
    ```ts
    export class MyStorePage {
      routeContext = resolve(IRouteContext);
    
      binding() {
        if (someLogic()) {
          this.routeContext.navigate('/somepath');
        }
      }
    }
    ```
    Or another form where we can specify the page component and have more hooks:
    ```html
    <au-viewport path="/store" component="my-store-page">
    ```
    ```ts
    export class MyStorePage {
      canLoad(context) {
        if (someLogic()) return '/somePath';
      }
    ```
    I think it can be made sufficient, but the starting point of this is to have the super simple plain way to do routing instead.

- For  `Class-less routed-components`

    From the nesting example, you can see that complexity/business is always manageable. With "class-less" (it can be a class, as per my `component="my-store-page"` example above), we have more choices: it can stay simple until it can't be:
    - Before:
    ```html
    <au-viewport path="/store">
      ... load and load and load of logic
    </au-viewport>
    ```
    - After
    ```html
    <au-viewport path="/store">
      <my-store-view>
    </au-viewport>
    ```
    it often takes a long time until you hit that concentration of logic though. Having this simplicity with the ability to convert is beneficial.

- For "guarding"

    > Employ a canLoad hook. If you have meant that, then I would like to point out that that adds one more point for the leaky encapsulation 

    I think I'm don't know exactly where you meant the `canLoad` hook to me, so maybe I'll just add an example
    ```html
    <au-viewport load.bind="..." can-load.bind="() => ..." can-unload.bind="" unload.bind="...">
    ```
    I didn't think of these hook originally since I just wanted something small and quick. This example may be even better and can help remove the need for `component="my-store-page"` just to have the ability to run some hook.

    > ... route-parameter constraint...

     I think I don't understand all of what you saw, so I'll just add an example of what I think can be fitting for route parameter constraint:
    ```html
        <au-viewport path="/store/:storeId{{^\\d+$}}">
    ```
    It looks just as good to me 🐱

- For conditional routing

    > I think that instead of using if.bind we can stick to the router-hooks, as those are actually meant for this purpose. It seems to me like reinventing the wheel.

    Originally I didn't think of `can-load.bind`/`can-unload.bind` etc... so if was chosen. Though I think if it's in the template, there's no need to restrain this, unless it adds too much to the code complexity, or too hard to implement. If it's just a few lines of code, the benefits is good I think.

- For configured routing integration

    > The RFC does not mention how the new pattern with play with the current practice. Think of a hybrid app, using configured routing (current practice) as well as the declarative routing.

    I'm not sure if they can play together, maybe they can since `<au-viewport>` **is the routed component**, we are just inlining the template of it. **Though** if we are to do it, I'd prefer a new name so it doesn't get confusing: i.e when is a viewport a viewport and when is a viewport a routed-component

- For testing

    You can create a mock router that upon app start, it will jump straight into the intended route, I don't think there's a complication in testing. For all the hooks ,they are just function so you can assert them like 1 + 1 = 2.
    For e2e, it'll just be the same with traditional e2e setup

- For implementation

    > For the implementation part, I think ideally we should have something like a router-plugin that implements the new pattern and

    Yeah, if we are to have a `<au-route/>` as an inline-able routed-component, it'll need not to be a plugin, can be just a built-in feature.

---

Thanks for the critics  @Sayan751 it helped me think a bit clearer 👍

**Comment by Sayan751** [comment]
Thanks @bigopon for the detailed response. I generally agree that for smaller apps, this feature might be tempting. Moreover, I would like to clarify on the following aspect to avoid any misunderstanding.

### Defining routing configurations on <au-viewport>

IMO the responsibility of an `au-viewport` is not to define routing configuration, rather to show/host a routed component. Hence, it does not quite sit right with me, as _my interpretation_ of that is that there are several viewports, and each one of those is responsible to show one (or more) particular route(s). Probably we need to define a new CE (`au-route`?) for that.

Or does the proposal really mean the existence of multiple (sibling) viewports, and each one gets activated based on the route and the condition (`if.bind`)?

### Binding to the routed components

>  If we were allowed to say I'm used to ..,, I'd say I'm used to plain and simple template with fewer to no indirections

I was only trying explicitly say that the pattern is unusual to _me_, and it was in no way an attempt to project my own design choices as the _general_ one.

According to my personal mental model, the difference between a regular custom element and a routed component is that a routed component in itself a "page" (inside a webapp) and is capable of getting all the data needed using various "services" during the activation phase, whereas a regular CE requires data from the host component to do, what it is intended to do. By leaky encapsulation, I meant that this new pattern does not fit my mental model of the routed components. Again, it does not mean that nobody should do it.

> May I know sufficient in which area? If we are talking about routing logic, maybe another perspective is to ask whether we want to scatter the routing logic all over the places? Maybe logic of routing should be mishmash-ed into the app component, at least until it's not suitable.

Regarding this one of your comments, I think we were probably referring to two different things. The point from me regarding the "mishmash of logic" refers to binding data as opposed to the routing logic, to the routed view-model. I am assuming that when you spoke about binding in your original post, you did not mean binding some routing logic to the components.

### Route gurading

Sorry about my last comment regarding this, as I missed the question :P I was trying to mean that your example in the original post is not clear to me (hence I listed all three alternatives and meant to enquire which one did you refer); as in what you meant by `guard` in those examples.

> ```html
> <h1>Items management</h1>
> <au-viewport path="/store">...</au-viewport>
> 
> <au-viewport path="/warehouse" guard.bind="(path, params, ...etc) => ...">
> <au-viewport path="/warehouse" guard.bind="state.isWarehouseManager">
> ```

I would appreciate it if you can elaborate it a little bit.

### Integration with configured routing

> I'm not sure if they can play together

If those cannot co-exist, the RFC should explicitly mention that as out-of-scope.

**Comment by bigopon** [comment]
- For au-viewport vs au-route:
    Yes I think we likely need to have a new au-route to express an inlined routed-component and leave au-viewport alone

- For multi viewport part:
    > Or does the proposal really mean the existence of multiple (sibling) viewports, and each one gets activated based on the route and the condition (if.bind)?
    
    I didn't think of multi viewport feature at all 😓, I guess when we specify `path="..."`, it can imply `path.match("...")` rather than `path === ` so that multiple viewport feature can be supported

- For page vs non page mental model + route guarding
    >  ... you did not mean binding some routing logic to the components...
    > ... route guard with `guard.bind...`
    
    One question that I think is suitable for looking at this is: Should the routed component itself knows about its loading logic, or should the component that owns it know that? If I'm able to agree with the later, it can be simpler for me to accept this inline-able routed component.
    I think I did not understand what I want clearly and did a poor job investigating what I missed. It only became a lot clearer to me after your questions in the first reply. So I originally thought of something a mix between a viewport, and a component that has the ability to express routing template. Now I think I can say that it's a inlined routed component that this RFC is about. With that, please disregard the part about `binding...` and `guard.bind` i mentioned previously. It should only be about `loading.bind...`/`canLoad.bind...`/etc...

> For integration

I think we will know clearer once there's some POC

**Comment by gavinaiken** [comment]
I don't mean to be negative here, but I'd really like to see Aurelia 2 transition to a production release version with the routing capabilities it already has, i.e. for the dev team to focus for now as much as possible on getting any remaining issues resolved which are preventing release, rather than adding new capabilities while it is in beta. A new router, or new capabilities in an existing router, could be added as soon as the release is complete and once any initial release issues have been resolved. I'm sure there's a large community of Aurelia 1 users out there with apps in production (like me!) who are waiting for final release of 2 before attempting an upgrade of existing apps. The sooner the final release comes, the better, from my point of view.

**Comment by dkent600** [comment]
Agreed, that a priority of this conversation should be to determine what extent Router and Router-Lite would be affected by this change.  Ideally not at all, then can go ahead and release AU without this new feature (which to me seems really cool) thus minimizing the risk of releasing this new feature later, and being clear that a later release would not break existing code.

If on the other hand there would be implications, then clearly enumerate them here.

**Comment by bigopon** [comment]
Thanks @gavinaiken @dkent600 , this will not affect our RC. I already have a POC that I think to implement it shouldn't have an effect on any existing routers. For RC, the aim is still beta 23-24.

**Comment by dkent600** [comment]
BTW, noting that router and route-lite have some differences in architecture, I wonder if this proposal would be biased towards one design or the other?

**Comment by bigopon** [comment]
@dkent600 it's not biased towards any, it's actually just a "pseudo router" (my name for it), as it's aimed for a very stripped down router + plain templating. This idea comes from my time working with react router, so you can say it's based on react router, but in our templating context.

**Comment by bigopon** [comment]
I've published a package based on this RFC at https://www.npmjs.com/package/aurelia-router-html

documentation is at https://aurelia-router-html.netlify.app/

This will not be at core repo package, at least not until it's been proven, battle tested.

**Comment by bigopon** [comment]
Closing this as this is now an experimental feature implementted in the package above. If it ever becomes stable enough we can bring it to the core. 🤞
